### DISCUSSION TITLE
model validation question

### DISCUSSION THREAD

**Discussion by rkdtmdals7710** [ultralytics/ultralytics#21308] [Q&A]
What do I do with the model.val when I have only the test data set among the train val test data sets?

The yaml file says there's no training data

I wanna know how to write data.yaml file about train : because I have only validation dataset

I think confidence will be the most efficient when you get the highest f1 score in a particular confidence, but when you get an f1 score, does iou basically set the threshold to 0.5? Or what do you set it to and get it?

+ If I validate with custom data.yaml, when I change class num of label or names of data.yaml, I cannot change the result. And val_batch0_labels.jpg show same predict result. How can I fix it?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rkdtmdals7710, thank you for reaching out to Ultralytics 🚀! For new users, we recommend checking out the [Docs](https://docs.ultralytics.com/) where you'll find detailed [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many frequently asked questions are already covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate.

If this is a custom training ❓ Question, please include as much detail as possible—such as your YAML configuration, dataset structure, and any relevant logs. Also, review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community wherever you prefer! For real-time chat, check out [Discord](https://discord.com/invite/ultralytics) 🎧. For longer-form discussions, see [Discourse](https://community.ultralytics.com/), or connect with other users on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

Upgrade to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to check if your issue persists:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests check all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response 🤖. An Ultralytics engineer will review your question and provide further assistance soon!

**Reply by glenn-jocher** [reply]
For validation-only scenarios, you can structure your data.yaml with just the validation path:

```yaml
val: path/to/your/test/data
nc: 3  # number of classes
names: ['class1', 'class2', 'class3']
```

Then run: `model.val(data='your_data.yaml')`

For F1 score calculation, the default IoU threshold is 0.5, but you can adjust it using the `iou` parameter: `model.val(data='data.yaml', iou=0.6)`. The confidence threshold that maximizes F1 score is often found through validation sweeps.

Regarding your custom data.yaml issue - ensure you're loading a model compatible with your class count, or use `model.val(data='data.yaml', verbose=True)` to see if your class names are being recognized. The model architecture must match your number of classes for proper evaluation.

**Comment by Y-T-G** [comment]
You still need to have `train:` key in the YAML. Just leave it blank.
